llmfoundry/utils/registry_utils.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import functools
+import importlib.util
+import os
+from pathlib import Path
+from types import ModuleType
+from typing import (Any, Callable, Dict, Generic, Optional, Sequence, Type,
+                    TypeVar, Union)
+
+import catalogue
+
+__all__ = ['TypedRegistry', 'create', 'builder']
+
+T = TypeVar('T')
+S = TypeVar('S')
+
+
+class TypedRegistry(catalogue.Registry, Generic[T]):
+    """A thin wrapper around catalogue.Registry to add static typing."""
+
+    def __init__(self,
+                 namespace: Sequence[str],
+                 entry_points: bool = False,
+                 description: str = '') -> None:
+        super().__init__(namespace, entry_points=entry_points)
+
+        self.description = description
+
+    def __call__(self, name: str, func: Optional[T] = None) -> Callable[[T], T]:
+        return super().__call__(name, func)
+
+    def register(self, name: str, *, func: Optional[T] = None) -> T:
+        return super().register(name, func=func)
+
+    def get(self, name: str) -> T:
+        return super().get(name)
+
+    def get_all(self) -> Dict[str, T]:
+        return super().get_all()
+
+    def get_entry_point(self, name: str, default: Optional[T] = None) -> T:
+        return super().get_entry_point(name, default=default)
+
+    def get_entry_points(self) -> Dict[str, T]:
+        return super().get_entry_points()
+
+
+def create(
```

Comment:
how about `create_registry`?

Or you can make it static method so the usage is `TypedRegistry.create(...)`

Reply:
yup, changed to `create_registry`